src/main/scala/mesosphere/marathon/core/readiness/impl/ReadinessCheckExecutorImpl.scala

```diff
@@ -0,0 +1,80 @@
+package mesosphere.marathon.core.readiness.impl
+
+import akka.actor.ActorSystem
+import akka.util.Timeout
+import mesosphere.marathon.core.readiness.ReadinessCheckExecutor.ReadinessCheckSpec
+import mesosphere.marathon.core.readiness.{ HttpResponse, ReadinessCheckExecutor, ReadinessCheckResult }
+import org.slf4j.LoggerFactory
+import rx.lang.scala.Observable
+import spray.client.pipelining._
+import spray.http.HttpHeaders.`Content-Type`
+import spray.http.{ HttpResponse => SprayHttpResponse, _ }
+
+import scala.concurrent.Future
+import scala.concurrent.duration.FiniteDuration
+import scala.util.Success
+import scala.util.control.NonFatal
+
+/**
+  * A Spray-based implementation of a ReadinessCheckExecutor.
+  */
+private[readiness] class ReadinessCheckExecutorImpl(implicit actorSystem: ActorSystem)
+    extends ReadinessCheckExecutor {
+
+  import scala.concurrent.ExecutionContext.Implicits.global
+  private[this] val log = LoggerFactory.getLogger(getClass)
+
+  override def execute(readinessCheckSpec: ReadinessCheckSpec): Observable[ReadinessCheckResult] = {
+    def singleCheck(): Future[ReadinessCheckResult] = executeSingleCheck(readinessCheckSpec)
+
+    intervalObservable(readinessCheckSpec.interval)
+      .scan(singleCheck()) {
```

Comment:
Exactly. Otherwise there is a slight race. If the next checks are initiated before the "unsubscribe" signal reaches the prior steps, we would perform unnecessary checks.